File: macos/BorderlessMouse/Sources/Security/PairingSecurity.swift
import CryptoKit
import Foundation
import Security

/// Kod parowania jest tylko czytelną reprezentacją 128-bitowego sekretu.
/// Myślniki i wielkość liter nie mają znaczenia przy wpisywaniu kodu.
enum PairingCodeCodec {
    private static let alphabet = Array("ABCDEFGHIJKLMNOPQRSTUVWXYZ234567".utf8)

    static func encode(_ data: Data) -> String {
        var output = [UInt8]()
        var accumulator: UInt32 = 0
        var bits = 0
        for byte in data {
            accumulator = (accumulator << 8) | UInt32(byte)
            bits += 8
            while bits >= 5 {
                bits -= 5
                output.append(alphabet[Int((accumulator >> UInt32(bits)) & 0x1F)])
            }
        }
        if bits > 0 {
            output.append(alphabet[Int((accumulator << UInt32(5 - bits)) & 0x1F)])
        }
        let raw = String(decoding: output, as: UTF8.self)
        return stride(from: 0, to: raw.count, by: 5).map { start in
            let a = raw.index(raw.startIndex, offsetBy: start)
            let b = raw.index(a, offsetBy: min(5, raw.count - start))
            return String(raw[a..<b])
        }.joined(separator: "-")
    }

    static func decode(_ code: String) -> Data? {
        let normalized = code.uppercased().filter { $0.isLetter || $0.isNumber }
        // 128 bitów daje dokładnie 26 znaków Base32. Dwa najmłodsze bity
        // ostatniego znaku są dopełnieniem i muszą pozostać zerowe.
        guard normalized.utf8.count == 26,
              let last = normalized.utf8.last,
              let lastIndex = alphabet.firstIndex(of: last),
              lastIndex & 0x03 == 0 else { return nil }
        var lookup = [UInt8: UInt8]()
        for (index, byte) in alphabet.enumerated() { lookup[byte] = UInt8(index) }
        var output = [UInt8]()
        var accumulator: UInt32 = 0
        var bits = 0
        for byte in normalized.utf8 {
            guard let value = lookup[byte] else { return nil }
            accumulator = (accumulator << 5) | UInt32(value)
            bits += 5
            if bits >= 8 {
                bits -= 8
                output.append(UInt8((accumulator >> UInt32(bits)) & 0xFF))
            }
        }
        guard output.count == 16 else { return nil }
        return Data(output)
    }
}

/// Sekret parowania nigdy nie trafia do UserDefaults ani logów.
final class PairingKeyStore {
    static let shared = PairingKeyStore()

    private let service = "com.borderlessmouse.pairing"
    private let account = "primary"
    private(set) var storageError: String?
    private var storedKey: Data

    private init() {
        if let existing = Self.read(service: service, account: account), existing.count == 16 {
            storedKey = existing
        } else {
            storedKey = Self.randomKey()
            storageError = Self.write(storedKey, service: service, account: account)
        }
    }

    var key: Data { storedKey }
    var displayCode: String { PairingCodeCodec.encode(storedKey) }

    @discardableResult
    func regenerate() -> String? {
        let next = Self.randomKey()
        if let error = Self.write(next, service: service, account: account) {
            storageError = error
            return error
        }
        storedKey = next
        storageError = nil
        return nil
    }

    private static func randomKey() -> Data {
        let count = 16
        var data = Data(count: count)
        let status = data.withUnsafeMutableBytes { bytes in
            SecRandomCopyBytes(kSecRandomDefault, count, bytes.baseAddress!)
        }
        precondition(status == errSecSuccess, "Nie można wygenerować bezpiecznego kodu parowania")
        return data
    }

    private static func read(service: String, account: String) -> Data? {
        let query: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: account,
            kSecReturnData: true,
            kSecMatchLimit: kSecMatchLimitOne,
        ]
        var item: CFTypeRef?
        guard SecItemCopyMatching(query as CFDictionary, &item) == errSecSuccess else { return nil }
        return item as? Data
    }

    private static func write(_ data: Data, service: String, account: String) -> String? {
        let identity: [CFString: Any] = [
            kSecClass: kSecClassGenericPassword,
            kSecAttrService: service,
            kSecAttrAccount: account,
        ]
        let update: [CFString: Any] = [
            kSecValueData: data,
            kSecAttrAccessible: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly,
        ]
        let updateStatus = SecItemUpdate(identity as CFDictionary, update as CFDictionary)
        if updateStatus == errSecSuccess { return nil }
        guard updateStatus == errSecItemNotFound else { return "Keychain: błąd \(updateStatus)" }
        var item = identity
        item.merge(update) { _, new in new }
        let status = SecItemAdd(item as CFDictionary, nil)
        return status == errSecSuccess ? nil : "Keychain: błąd \(status)"
    }
}

enum ControlCrypto {
    static let nonceBytes = 16
    static let proofBytes = 32
    static let tagBytes = 16
    static let counterBytes = 8

    static func randomNonce() -> Data {
        let count = nonceBytes
        var data = Data(count: count)
        let status = data.withUnsafeMutableBytes { bytes in
            SecRandomCopyBytes(kSecRandomDefault, count, bytes.baseAddress!)
        }
        precondition(status == errSecSuccess, "Nie można wygenerować nonce")
        return data
    }

    static func proof(secret: Data, role: String, clientNonce: Data, serverNonce: Data) -> Data {
        var message = Data("BorderlessMouse/v2/\(role)".utf8)
        message.append(clientNonce)
        message.append(serverNonce)
        let code = HMAC<SHA256>.authenticationCode(for: message, using: SymmetricKey(data: secret))
        return Data(code)
    }

    static func constantTimeEqual(_ lhs: Data, _ rhs: Data) -> Bool {
        guard lhs.count == rhs.count else { return false }
        var difference: UInt8 = 0
        for (a, b) in zip(lhs, rhs) { difference |= a ^ b }
        return difference == 0
    }

    static func derive(secret: Data, salt: Data, info: String, bytes: Int = 32) -> Data {
        let key = HKDF<SHA256>.deriveKey(inputKeyMaterial: SymmetricKey(data: secret),
                                         salt: salt,
                                         info: Data(info.utf8),
                                         outputByteCount: bytes)
        return key.withUnsafeBytes { Data($0) }
    }
}

/// Szyfruje kompletne ramki protokołu. TCP zapewnia kolejność, dlatego licznik
/// musi być ściśle rosnący; powtórzona lub przestawiona ramka kończy sesję.
final class SecureSession {
    enum Role { case client, server }

    private let sendKey: SymmetricKey
    private let receiveKey: SymmetricKey
    let audioKey: Data
    let audioSessionID: UInt64
    private var sendCounter: UInt64 = 0
    private var receiveCounter: UInt64?
    private let lock = NSLock()

    init(secret: Data, clientNonce: Data, serverNonce: Data, role: Role) {
        var salt = Data()
        salt.append(clientNonce)
        salt.append(serverNonce)
        let clientToServer = ControlCrypto.derive(secret: secret, salt: salt, info: "BorderlessMouse/v2/control/client-to-server")
        let serverToClient = ControlCrypto.derive(secret: secret, salt: salt, info: "BorderlessMouse/v2/control/server-to-client")
        switch role {
        case .client:
            sendKey = SymmetricKey(data: clientToServer)
            receiveKey = SymmetricKey(data: serverToClient)
        case .server:
            sendKey = SymmetricKey(data: serverToClient)
            receiveKey = SymmetricKey(data: clientToServer)
        }
        audioKey = ControlCrypto.derive(secret: secret, salt: salt, info: "BorderlessMouse/v2/audio/server-to-client")
        let idData = ControlCrypto.derive(secret: secret, salt: salt, info: "BorderlessMouse/v2/audio/session-id", bytes: 8)
        audioSessionID = idData.withUnsafeBytes { raw in
            raw.loadUnaligned(as: UInt64.self).littleEndian
        }
    }

    func seal(_ plaintext: Data) -> Data? {
        lock.lock()
        defer { lock.unlock() }
        guard sendCounter != UInt64.max else { return nil }
        let counter = sendCounter
        sendCounter &+= 1
        let counterData = Self.littleEndian(counter)
        do {
            let sealed = try AES.GCM.seal(plaintext,
                                          using: sendKey,
                                          nonce: try AES.GCM.Nonce(data: Self.nonce(counter)),
                                          authenticating: Self.aad(counterData))
            var output = counterData
            output.append(sealed.ciphertext)
            output.append(sealed.tag)
            return output
        } catch {
            return nil
        }
    }

    func open(_ envelope: Data) -> Data? {
        lock.lock()
        defer { lock.unlock() }
        guard envelope.count >= ControlCrypto.counterBytes + ControlCrypto.tagBytes else { return nil }
        let counterData = envelope.prefix(ControlCrypto.counterBytes)
        let counter = counterData.withUnsafeBytes { $0.loadUnaligned(as: UInt64.self).littleEndian }
        if let previous = receiveCounter, counter <= previous { return nil }
        let ciphertext = envelope.dropFirst(ControlCrypto.counterBytes).dropLast(ControlCrypto.tagBytes)
        let tag = envelope.suffix(ControlCrypto.tagBytes)
        do {
            let box = try AES.GCM.SealedBox(nonce: AES.GCM.Nonce(data: Self.nonce(counter)),
                                            ciphertext: ciphertext,
                                            tag: tag)
            let clear = try AES.GCM.open(box, using: receiveKey, authenticating: Self.aad(Data(counterData)))
            receiveCounter = counter
            return clear
        } catch {
            return nil
        }
    }

    private static func nonce(_ counter: UInt64) -> Data {
        Data([0x42, 0x4C, 0x4D, 0x32]) + littleEndian(counter)
    }

    private static func aad(_ counter: Data) -> Data {
        Data("BLM2".utf8) + counter
    }

    private static func littleEndian(_ value: UInt64) -> Data {
        var v = value.littleEndian
        return withUnsafeBytes(of: &v) { Data($0) }
    }
}

File: windows/BorderlessMouse/Security/PairingSecurity.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace BorderlessMouse.Security;

public static class PairingCodeCodec
{
    private const string Alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";

    public static string Encode(ReadOnlySpan<byte> data)
    {
        var output = new StringBuilder((data.Length * 8 + 4) / 5);
        uint accumulator = 0;
        var bits = 0;
        foreach (var value in data)
        {
            accumulator = (accumulator << 8) | value;
            bits += 8;
            while (bits >= 5)
            {
                bits -= 5;
                output.Append(Alphabet[(int)((accumulator >> bits) & 0x1F)]);
            }
        }
        if (bits > 0) output.Append(Alphabet[(int)((accumulator << (5 - bits)) & 0x1F)]);
        var raw = output.ToString();
        return string.Join('-', Enumerable.Range(0, (raw.Length + 4) / 5)
            .Select(i => raw.Substring(i * 5, Math.Min(5, raw.Length - i * 5))));
    }

    public static bool TryDecode(string? code, out byte[] key)
    {
        key = Array.Empty<byte>();
        if (string.IsNullOrWhiteSpace(code)) return false;
        var normalized = new string(code.Where(char.IsLetterOrDigit).Select(char.ToUpperInvariant).ToArray());
        if (normalized.Length != 26) return false;
        var lastValue = Alphabet.IndexOf(normalized[^1]);
        if (lastValue < 0 || (lastValue & 0x03) != 0) return false;
        var output = new List<byte>();
        uint accumulator = 0;
        var bits = 0;
        foreach (var ch in normalized)
        {
            var value = Alphabet.IndexOf(ch);
            if (value < 0) return false;
            accumulator = (accumulator << 5) | (uint)value;
            bits += 5;
            if (bits >= 8)
            {
                bits -= 8;
                output.Add((byte)((accumulator >> bits) & 0xFF));
            }
        }
        if (output.Count != 16) return false;
        key = output.ToArray();
        return true;
    }
}

/// <summary>Przechowuje sekret parowania zaszyfrowany dla bieżącego konta Windows (DPAPI).</summary>
public static class PairingKeyStore
{
    private static readonly byte[] Entropy = Encoding.UTF8.GetBytes("BorderlessMouse/v2/pairing-key");
    public static string FilePath => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "BorderlessMouse", "pairing.key");

    public static byte[]? Load()
    {
        if (!OperatingSystem.IsWindows() || !File.Exists(FilePath)) return null;
        try
        {
            var protectedBytes = File.ReadAllBytes(FilePath);
            var key = ProtectedData.Unprotect(protectedBytes, Entropy, DataProtectionScope.CurrentUser);
            if (key.Length == 16) return key;
            CryptographicOperations.ZeroMemory(key);
            return null;
        }
        catch { return null; }
    }

    public static bool SaveCode(string code)
    {
        if (!OperatingSystem.IsWindows() || !PairingCodeCodec.TryDecode(code, out var key)) return false;
        Directory.CreateDirectory(Path.GetDirectoryName(FilePath)!);
        byte[]? protectedBytes = null;
        string? temporary = null;
        try
        {
            protectedBytes = ProtectedData.Protect(key, Entropy, DataProtectionScope.CurrentUser);
            temporary = FilePath + "." + Guid.NewGuid().ToString("N") + ".tmp";
            File.WriteAllBytes(temporary, protectedBytes);
            File.Move(temporary, FilePath, true);
            return true;
        }
        finally
        {
            CryptographicOperations.ZeroMemory(key);
            if (protectedBytes is not null) CryptographicOperations.ZeroMemory(protectedBytes);
            try { if (temporary is not null && File.Exists(temporary)) File.Delete(temporary); } catch { }
        }
    }

    public static void Clear()
    {
        if (File.Exists(FilePath)) File.Delete(FilePath);
    }
}

public static class ControlCrypto
{
    public const int NonceBytes = 16;
    public const int ProofBytes = 32;
    public const int TagBytes = 16;
    public const int CounterBytes = 8;

    public static byte[] RandomNonce() => RandomNumberGenerator.GetBytes(NonceBytes);

    public static byte[] Proof(ReadOnlySpan<byte> secret, string role, ReadOnlySpan<byte> clientNonce, ReadOnlySpan<byte> serverNonce)
    {
        var prefix = Encoding.UTF8.GetBytes($"BorderlessMouse/v2/{role}");
        var message = new byte[prefix.Length + clientNonce.Length + serverNonce.Length];
        prefix.CopyTo(message, 0);
        clientNonce.CopyTo(message.AsSpan(prefix.Length));
        serverNonce.CopyTo(message.AsSpan(prefix.Length + clientNonce.Length));
        return HMACSHA256.HashData(secret, message);
    }

    public static byte[] Derive(ReadOnlySpan<byte> secret, ReadOnlySpan<byte> salt, string info, int length = 32)
    {
        var prk = HMACSHA256.HashData(salt, secret);
        var output = new byte[length];
        var previous = Array.Empty<byte>();
        var offset = 0;
        byte counter = 1;
        var infoBytes = Encoding.UTF8.GetBytes(info);
        while (offset < length)
        {
            var input = new byte[previous.Length + infoBytes.Length + 1];
            previous.CopyTo(input, 0);
            infoBytes.CopyTo(input, previous.Length);
            input[^1] = counter++;
            previous = HMACSHA256.HashData(prk, input);
            var count = Math.Min(previous.Length, length - offset);
            previous.AsSpan(0, count).CopyTo(output.AsSpan(offset));
            offset += count;
        }
        CryptographicOperations.ZeroMemory(prk);
        return output;
    }
}

/// <summary>Dwukierunkowa sesja AES-256-GCM dla kompletnych ramek protokołu.</summary>
public sealed class SecureSession : IDisposable
{
    public enum SessionRole { Client, Server }

    private readonly byte[] _sendKey;
    private readonly byte[] _receiveKey;
    private readonly object _sendGate = new();
    private readonly object _receiveGate = new();
    private ulong _sendCounter;
    private ulong? _receiveCounter;
    private bool _disposed;

    public byte[] AudioKey { get; }
    public ulong AudioSessionId { get; }

    public SecureSession(ReadOnlySpan<byte> secret, ReadOnlySpan<byte> clientNonce, ReadOnlySpan<byte> serverNonce, SessionRole role)
    {
        var salt = new byte[clientNonce.Length + serverNonce.Length];
        clientNonce.CopyTo(salt);
        serverNonce.CopyTo(salt.AsSpan(clientNonce.Length));
        var c2s = ControlCrypto.Derive(secret, salt, "BorderlessMouse/v2/control/client-to-server");
        var s2c = ControlCrypto.Derive(secret, salt, "BorderlessMouse/v2/control/server-to-client");
        (_sendKey, _receiveKey) = role == SessionRole.Client ? (c2s, s2c) : (s2c, c2s);
        AudioKey = ControlCrypto.Derive(secret, salt, "BorderlessMouse/v2/audio/server-to-client");
        var session = ControlCrypto.Derive(secret, salt, "BorderlessMouse/v2/audio/session-id", 8);
        AudioSessionId = BinaryPrimitives.ReadUInt64LittleEndian(session);
        CryptographicOperations.ZeroMemory(session);
    }

    public byte[]? Seal(ReadOnlySpan<byte> plaintext)
    {
        lock (_sendGate)
        {
            if (_disposed) return null;
            if (_sendCounter == ulong.MaxValue) return null;
            var counter = _sendCounter++;
            var output = new byte[ControlCrypto.CounterBytes + plaintext.Length + ControlCrypto.TagBytes];
            BinaryPrimitives.WriteUInt64LittleEndian(output, counter);
            var nonce = Nonce(counter);
            var aad = Aad(output.AsSpan(0, ControlCrypto.CounterBytes));
            using var aes = new AesGcm(_sendKey, ControlCrypto.TagBytes);
            aes.Encrypt(nonce, plaintext,
                output.AsSpan(ControlCrypto.CounterBytes, plaintext.Length),
                output.AsSpan(ControlCrypto.CounterBytes + plaintext.Length, ControlCrypto.TagBytes), aad);
            return output;
        }
    }

    public byte[]? Open(ReadOnlySpan<byte> envelope)
    {
        lock (_receiveGate)
        {
            if (_disposed) return null;
            if (envelope.Length < ControlCrypto.CounterBytes + ControlCrypto.TagBytes) return null;
            var counter = BinaryPrimitives.ReadUInt64LittleEndian(envelope);
            if (_receiveCounter is { } previous && counter <= previous) return null;
            var clearLength = envelope.Length - ControlCrypto.CounterBytes - ControlCrypto.TagBytes;
            var plaintext = new byte[clearLength];
            try
            {
                using var aes = new AesGcm(_receiveKey, ControlCrypto.TagBytes);
                aes.Decrypt(Nonce(counter),
                    envelope.Slice(ControlCrypto.CounterBytes, clearLength),
                    envelope.Slice(ControlCrypto.CounterBytes + clearLength, ControlCrypto.TagBytes),
                    plaintext, Aad(envelope[..ControlCrypto.CounterBytes]));
                _receiveCounter = counter;
                return plaintext;
            }
            catch (CryptographicException)
            {
                CryptographicOperations.ZeroMemory(plaintext);
                return null;
            }
        }
    }

    private static byte[] Nonce(ulong counter)
    {
        var nonce = new byte[12] { 0x42, 0x4C, 0x4D, 0x32, 0, 0, 0, 0, 0, 0, 0, 0 };
        BinaryPrimitives.WriteUInt64LittleEndian(nonce.AsSpan(4), counter);
        return nonce;
    }

    private static byte[] Aad(ReadOnlySpan<byte> counter)
    {
        var aad = new byte[4 + ControlCrypto.CounterBytes];
        "BLM2"u8.CopyTo(aad);
        counter.CopyTo(aad.AsSpan(4));
        return aad;
    }

    public void Dispose()
    {
        lock (_sendGate)
        lock (_receiveGate)
        {
            if (_disposed) return;
            _disposed = true;
            CryptographicOperations.ZeroMemory(_sendKey);
            CryptographicOperations.ZeroMemory(_receiveKey);
            CryptographicOperations.ZeroMemory(AudioKey);
        }
    }
}
